README.md

# OmicSense

<!-- badges: start -->
<!-- badges: end -->

The goal of OmicSense is to provide the functions to construct a prediction model of environments using noisy omics data linked with the environments based on OmicSense algorithm. 

## Installation
You can install the development version of OmicSense from [GitHub](https://github.com/takakoizumi/OmicSense) with:    
#install.packages("devtools")   
devtools::install_github("takakoizumi/OmicSense") 

## Example    
install.packages("OmicSense")   
library(OmicSense)    
# basic example code   
data(Pinus)   
x.train <- Pinus[[1]]   
x.test <- Pinus[[2]]    
y <- Pinus[[3]]   
cor(y, os.pred(x.train, y, newx = x.test, n.pred = 100))


NEWS.md
---
title: "NEWS"
author: "Takahiko Koizumi"
date: "22/09/2023"
output: html_document
---

## OmicSense 0.2.0




tests/testthat.R
library(testthat)
library(OmicSense)

test_check("OmicSense")


tests/testthat/test_OmicSense.R
test_that("sort", {

  y <- runif(30)
  x <- matrix(runif(30 * 30), 30, 30)

  expect_equal(nrow(os.sort(x, y)), 30)
  expect_equal(ncol(os.sort(x, y)), 30)
})


R/os.opt.R
##' Estimate the optimal number of predictors to construct OmicSense prediction model
##'
##' @title Estimate the optimal number of predictors to construct OmicSense prediction model
##' @param x A data matrix (row: samples, col: predictors).
##' @param y A vector of target value.
##' @param range A sequence of numbers of predictors to be tested for MAE calculation (default: 5:50).
##' @param method A string to specify the regression function for calculating R-squared values.
##' "linear" (default), "quadratic" or "cubic" function can be specified.
##' @param rep The number of replications for each case set by range (default: 1).
##' @importFrom ggplot2 ggplot aes geom_point stat_function geom_hline geom_text theme element_rect element_blank
##' @importFrom stats rnorm nls
##' @return A sample-MAE curve
##' @examples
##' data(Pinus)
##' train <- os.clean(Pinus$train)
##' target <- Pinus$target
##' os.opt(train[1:10, ], target[1:10], range = 5:15)
##' @author Takahiko Koizumi
##' @export
os.opt <- function(x, y, range = 5:50, method = "linear", rep = 1) {
  npred <- MAE <- NULL

  degree <- switch(method,
                   "linear" = 1,
                   "quadratic" = 2,
                   "cubic" = 3,
                   stop("Select the methods from linear, quadratic, or cubic")
  )

  if(min(range) < 0){
    stop("<range> must not include a negative value")
  }else if(max(range) > ncol(x)){
    stop(paste("<range> must not exceed", ncol(x), sep = " "))
  }
  if(rep <= 0){
    stop("<rep> must be a natural number")
  }

  g <- range
  unit <- rep(NA, length(g) * rep)
  comp <- data.frame(
    npred = unit,
    MAE = unit
  )

  ## sort predictors in descending order of R2
  x.ps <- os.sort(x, y, method = method)
  x.ps <- x.ps[, 1:max(g)]
  message("Data prepared")

  ## calculate MAE
  total <- length(range) * rep
  perc <- trunc(10 * c(1:total) / total)

  counter <- 1
  proc <- perc[1]
  for(i in 1:length(g)){
    for(j in 1:rep){
      x.p <- x.ps[, 1:g[i]]
      x.p.test <- x.p
      ## add noise
      for(k in 1:g[i]){
        p <- x.p[, k]
        x.p.test[, k] <- rnorm(length(p), p, p/2)
      }

      ## run OmicSense
      pc <- os.pred(x.p, y, newx = x.p.test, method = method)
      comp[counter, ] <- c(g[i], mean(abs(y - pc)))

      if(proc != perc[counter]){
        message(paste((proc + 1) * 10, "% completed", sep = ""))
        proc <- perc[counter]
      }
      counter <- counter + 1
    }
  }

  ## estimate the minimum MAE
  fit <- nls(MAE ~ a + b * exp(- c * npred), start = c (a = 1, b = 1, c = 0.1), data = comp)
  a <- fit$m$getPars()[[1]]
  b <- fit$m$getPars()[[2]]
  c <- fit$m$getPars()[[3]]

  ggplot(comp, aes(x = npred, y = MAE)) +
    geom_point(aes(colour = "MAE", fill = "MAE"), shape = 21) +
    stat_function(fun = function(x) a + b * exp(- c * x)) +
    geom_hline(aes(yintercept = a), colour = "red", linetype = "dotted") +
    geom_text(aes(0, a, label = format(round(a, 3), nsmall = 3), vjust = - 1)) +
    xlab("Number of variables") +
    theme(panel.background = element_rect(fill = "transparent", colour = "black"),
          panel.grid = element_blank(),
          strip.background = element_blank(),
          legend.position = "none")
}





R/os.pca.R
##' Visualize predictors using principal coordinate analysis
##'
##' @title Visualize predictors using principal coordinate analysis
##' @param x A data matrix (row: samples, col: predictors).
##' @param y A vector of target value.
##' @param method A string to specify the regression function for calculating R-squared values.
##' "linear" (default), "quadratic" or "cubic" function can be specified.
##' @param n.pred The number of candidate predictors for prediction model to be indicated in a PCA plot (default: ncol(x)).
##' @param thresh The lower threshold of R-squared value to be indicated in a PCA plot (default: 0).
##' @param size The size of symbols in a PCA plot (default: 1).
##' @importFrom ggplot2 ggplot aes geom_point xlab ylab scale_fill_manual scale_fill_gradient theme element_rect element_blank
##' @importFrom stats lm princomp
##' @return A PCA plot
##' @examples
##' data(Pinus)
##' train <- os.clean(Pinus$train)
##' target <- Pinus$target
##' os.pca(train, target)
##' @author Takahiko Koizumi
##' @export
os.pca <- function(x, y, method = "linear", thresh = 0, n.pred = ncol(x), size = 1){
  PC1 <- PC2 <- R2 <- selection <- NULL

  pc <- princomp(t(scale(x)), cor = TRUE, scores = TRUE)
  coords <- pc$scores
  sd <- summary(pc)$sdev
  pc1 <- (sd ^ 2 / sum(sd ^ 2))[1]
  pc2 <- (sd ^ 2 / sum(sd ^ 2))[2]

  degree <- switch(method,
                   "linear" = 1,
                   "quadratic" = 2,
                   "cubic" = 3,
                   stop("Select the methods linear, quadratic, or cubic")
  )

  if(thresh > 0 & n.pred < ncol(x)){
    stop("Don't specify <n.pred> and <thresh> at a time")
  }
  if(thresh < 0 | thresh > 1){
    stop("<thresh> should be within the range of 0-1")
  }
  if(n.pred < 0){
    stop("<n.pred> should not be a negative value")
  }else if(n.pred > ncol(x)){
    stop(paste("<n.pred> must not exceed", ncol(x), sep = " "))
  }

  ## calculate R2 values
  r2 <- rep(NA, ncol(x))
  rc <- rep(NA, ncol(x))
  for(i in 1:ncol(x)){
    rs <- summary(lm(y ~ poly(x[, i], degree = degree, raw = TRUE)))
    r2[i] <- rs$r.squared
    rc[i] <- rs$coefficients[2, 1]
  }

  h <- data.frame(
    R2 = r2,
    Coef = rc,
    selection = rep(NA, ncol(x))
  )
  row.names(h) <- colnames(x)
  coords <- cbind(coords[, 1:2], h)
  colnames(coords) <- c("PC1", "PC2", "R2", "Coef", "selection")
  coords <- coords[order(r2, decreasing = FALSE), ]
  if(thresh > 0){
    coords[coords$R2 >= thresh, "selection"] <- "selected"
    coords[coords$R2 < thresh, "selection"] <- "not"
  }else if(n.pred < ncol(x)){
    coords[, "selection"] <- "selected"
    coords[1:(nrow(coords) - n.pred), "selection"] <- "not"
  }else{
    coords <- coords
  }

  if(thresh > 0 | n.pred < ncol(x)){
    ggplot(coords, aes(x = PC1, y = PC2)) +
      geom_point(aes(fill = selection), colour = "white", shape = 21, alpha = 0.5, stroke = 0.1, size = size) +
      scale_fill_manual(values = c("gray", "red")) +
      xlab(paste("PC1 (", format(round(pc1 * 100, digits = 1), nsmall = 1), "%)", sep = "")) +
      ylab(paste("PC2 (", format(round(pc2 * 100, digits = 1), nsmall = 1), "%)", sep = "")) +
      theme(panel.background = element_rect(fill = "transparent", colour = "black"),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            legend.position = "none")
  }else{
    coords[coords$Coef < 0, "R2"] <- -coords[coords$Coef < 0, "R2"]
    ggplot(coords, aes(x = PC1, y = PC2)) +
      geom_point(aes(fill = R2), colour = "white", shape = 21, alpha = 0.5, stroke = 0.1, size = size) +
      scale_fill_gradient(low = "gray", high = "red") +
      xlab(paste("PC1 (", format(round(pc1 * 100, digits = 1), nsmall = 1), "%)", sep = "")) +
      ylab(paste("PC2 (", format(round(pc2 * 100, digits = 1), nsmall = 1), "%)", sep = "")) +
      theme(panel.background = element_rect(fill = "transparent", colour = "black"),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            legend.position = "none")
  }
}





R/os.rank.R
##' Visualize R-squared value distribution in target-predictor relationship
##'
##' @title Visualize R-squared value distribution in target-predictor relationship
##' @param x A data matrix (row: samples, col: predictors).
##' @param y A vector of target value.
##' @param method A string to specify the regression function for calculating R-squared values.
##' "linear" (default), "quadratic" or "cubic" function can be specified.
##' @param n.pred The number of predictors to be leaved in prediction model (default: ncol(x)).
##' @param thresh The lower threshold of R-squared value to be leaved in prediction model (default: 0).
##' @param upper.xlim The upper limitation of x axis (i.e., the number of predictors) in the resulted figure (default: ncol(x)).
##' @importFrom ggplot2 ggplot aes geom_line geom_area xlim ylim xlab ylab
##' geom_hline geom_text geom_vline theme element_rect element_blank
##' @importFrom stats lm
##' @return A rank order plot
##' @examples
##' data(Pinus)
##' train <- os.clean(Pinus$train)
##' target <- Pinus$target
##' train <- os.sort(train, target)
##' os.rank(train, target)
##' @author Takahiko Koizumi
##' @export
os.rank <- function(x, y, method = "linear", thresh = 0, n.pred = ncol(x), upper.xlim = ncol(x)){
  R2 <- NULL
  degree <- switch(method,
                   "linear" = 1,
                   "quadratic" = 2,
                   "cubic" = 3,
                   stop("Select the <method> linear, quadratic, or cubic")
  )

  if(thresh > 0 & n.pred != ncol(x)){
    stop("Don't specify <n.pred> and <thresh> at a time")
  }
  if(thresh < 0 | thresh > 1){
    stop("<thresh> should be within the range of 0-1")
  }
  if(n.pred < 0){
    stop("<n.pred> should not be a negative value")
  }else if(n.pred > ncol(x)){
    stop(paste("<n.pred> must not exceed", ncol(x), sep = " "))
  }


  ## calculate R2 values
  result <- rep(NA, ncol(x))
  for(i in 1:ncol(x)){
    result[i] <- summary(lm(y ~ poly(x[, i], degree = degree, raw = TRUE)))$r.squared
  }

  h <- data.frame(
    rank = rep(NA, ncol(x)),
    R2 = result
  )

  ## sort predictors in descending order of R2
  h <- h[order(h$R2, decreasing = TRUE), ]
  h$rank <- 1:nrow(h)

  ## set the number of predictors to be used
  if(thresh != 0){
    n.pred <- length(result[result >= thresh])
  }

  if(thresh > 0){
    rankplot <- ggplot() +
      geom_line(data = h, aes(x = rank, y = R2), colour = "black") +
      geom_area(data = h, aes(x = rank, y = R2), fill = "gray") +
      geom_area(data = h[1:n.pred, ], aes(x = rank, y = R2), fill = "red") +
      xlim(0, upper.xlim) +
      ylim(0, 1) +
      xlab("Rank") +
      ylab("R-squared value") +
      geom_hline(aes(yintercept = thresh), colour = "black", linetype = "dotted") +
      geom_text(aes(0, thresh, label = format(round(thresh, 2), nsmall = 2), hjust = -0.2, vjust = 1.5)) +
      geom_vline(aes(xintercept = n.pred), colour = "black", linetype = "dotted") +
      geom_text(aes(n.pred, 0.03, label = n.pred, hjust = -0.5)) +
      theme(panel.background = element_rect(fill = "transparent", colour = "black"),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            legend.position = "none")
  }else{
    R2val <- h[n.pred, "R2"]
    rankplot <- ggplot() +
      geom_line(data = h, aes(x = rank, y = R2), colour = "black") +
      geom_area(data = h, aes(x = rank, y = R2), fill = "gray") +
      geom_area(data = h[1:n.pred, ], aes(x = rank, y = R2), fill = "red") +
      xlim(0, upper.xlim) +
      ylim(0, 1) +
      xlab("Rank") +
      ylab("R-squared value") +
      geom_hline(aes(yintercept = R2val), colour = "black", linetype = "dotted") +
      geom_text(aes(0, R2val, label = format(round(R2val, 2), nsmall = 2), hjust = -0.2, vjust = 1.5)) +
      geom_vline(aes(xintercept = n.pred), colour = "black", linetype = "dotted") +
      geom_text(aes(n.pred, 0.03, label = n.pred, hjust = -0.5)) +
      theme(panel.background = element_rect(fill = "transparent", colour = "black"),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            legend.position = "none")
  }
  suppressWarnings(rankplot)
}





R/os.sort.R
##' Sort and select predictors according to the strength of target-predictor relationship
##'
##' @title Sort and select predictors according to the strength of target-predictor relationship
##' @param x A data matrix (raw: samples, col: predictors).
##' @param y A vector of target value.
##' @param method A string to specify the regression function for calculating R-squared values.
##' "linear" (default), "quadratic" or "cubic" function can be specified.
##' @param n.pred The number of predictors to be leaved in prediction model (default: ncol(x)).
##' @param thresh The lower threshold of R-squared value to be leaved in prediction model (default: 1).
##' @importFrom stats lm
##' @return A data matrix (raw: samples, col: sorted predictors)
##' @examples
##' data(Pinus)
##' train <- os.clean(Pinus$train)
##' target <- Pinus$target
##' cor(target, train[, 1])
##'
##' train <- os.sort(train, target, thresh = 0.5)
##' cor(target, train[, 1])
##' @author Takahiko Koizumi
##' @export
os.sort <- function(x, y, method = "linear", n.pred = ncol(x), thresh = 1){
  degree <- switch(method,
                   "linear" = 1,
                   "quadratic" = 2,
                   "cubic" = 3,
                   stop("Select the <method> linear, quadratic, or cubic")
  )

  if(n.pred < ncol(x) & thresh < 1){
    stop("Don't specify <n.pred> and <thresh> at a time")
  }
  if(n.pred < 0){
    stop("<n.pred> should not be a negative value")
  }else if(n.pred > ncol(x)){
    stop(paste("<n.pred> must not exceed", ncol(x), sep = " "))
  }
  if(thresh < 0 | thresh > 1){
    stop("<thresh> should be within the range of 0-1")
  }

  ## calculate R2 values
  result <- rep(NA, ncol(x))
  for(i in 1:ncol(x)){
    result[i] <- summary(lm(y ~ poly(x[, i], degree = degree, raw = TRUE)))$r.squared
  }
  ## sort predictors in descending order of R2
  x <- x[, order(result, decreasing = TRUE)]

  ## extract predictors with higher R2
  if(n.pred <= ncol(x)){
    x <- x[, 1:n.pred]
  }else if(thresh < 1){
    x <- x[, 1:length(result[result >= thresh])]
  }else{
    x <- x
  }
  return(x)
}


R/os.clean.R
##' Clean data by eliminating predictors with many missing values
##'
##' @title Clean data by eliminating predictors with many missing values
##' @param x A data matrix (raw: samples, col: predictors).
##' @param missing A ratio of missing values in each column allowed to be remained in the data.
##' @param lowest The lowest value to be leaved in the data.
##' @return A data matrix (raw: samples, col: qualified predictors)
##' @examples
##' data(Pinus)
##' train.raw <- Pinus$train
##' ncol(train.raw)
##'
##' train <- os.clean(train.raw)
##' ncol(train)
##' @author Takahiko Koizumi
##' @export
os.clean <- function(x, missing = 0.1, lowest = 10){
  if(missing < 0 | missing > 1){
    stop("<missing> should be within the range of 0-1")
  }
  if(lowest < 0){
    stop("<lowest> should not be a negative value")
  }else if(lowest > ncol(x)){
    stop(paste("<lowest> must not exceed", max(x, na.rm = TRUE), sep = " "))
  }

  r <- function(z) length(z[z != 0]) / length(z)
  ## handle predictors with missing values
  x <- x[, apply(x, 2, r) >= (1 - missing)]
  ## handle predictors with low expression
  x <- x[, apply(x, 2, max) >= lowest]
  return(x)
}


R/os.pred.R
##' Construct and apply the OmicSense model with your own data
##'
##' @title Construct and apply the OmicSense model with your own data
##' @param x A data matrix (row: samples, col: predictors).
##' @param y A vector of target value.
##' @param newx A data matrix (row: samples, col: predictors).
##' @param method A string to specify the regression function for calculating R-squared values.
##' "linear" (default), "quadratic" or "cubic" function can be specified.
##' @param n.pred The number of candidate predictors to be leaved in prediction model (default: 30).
##' @param thresh The lower threshold of R-squared value to be leaved in prediction model (default: 0).
##' @importFrom stats lm dnorm residuals dnorm predict quantile
##' @importFrom utils capture.output
##' @importFrom kernlab gausspr
##' @return A vector of the environment in which the samples of newx were collected
##' @examples
##' data(Pinus)
##' train <- os.clean(Pinus$train)
##' test <- Pinus$test
##' test <- test[, colnames(train)]
##' target <- Pinus$target
##' cor(target, os.pred(train, target, newx = test, method = "cubic"))
##' @author Takahiko Koizumi
##' @export
os.pred <- function(x, y, newx = x, method = "linear", thresh = 0, n.pred = 0){
  degree <- switch(method,
                   "linear" = 1,
                   "quadratic" = 2,
                   "cubic" = 3,
                   "mix" = 0,
                   "gaussian" = -1,
                   stop("Select the methods linear, quadratic, cubic, mix, or gaussian")
  )

  if(thresh > 0 & n.pred > 0){
    stop("Don't specify <n.pred> and <thresh> at a time")
  }
  if(n.pred == 0){
    n.pred <- 30
  }


  ## definitions
  ## adjust errors
  adj <- function(xa, ya, xb){
    cf <- summary(lm(ya ~ xa))$coefficients
    xb * cf[2, 1] + cf[1, 1]
  }

  ## interquartile range
  quant <- function(xa){
    q <- quantile(xa)
    mn <- q[2] - 1.5 * (q[4] - q[2])
    mx <- q[4] + 1.5 * (q[4] - q[2])
    (1:length(xa))[xa >= mn & xa <= mx]
  }

  ## OmicSense with polynomial regression
  sense.poly <- function(x.train, y.train, newx,
                         method = method, n.pred = n.pred, threshold = thresh){

    degree <- switch(method,
                     "linear" = 1,
                     "quadratic" = 2,
                     "cubic" = 3,
                     stop("Select the methods linear, quadratic, or cubic")
    )

    ## calculate R2 values
    result <- rep(NA, ncol(x.train))
    for(i in 1:ncol(x.train)){
      result[i] <- summary(lm(y.train ~ poly(x.train[, i], degree = degree, raw = TRUE)))$r.squared
    }

    ## sort predictors in descending order of R2
    x.train <- x.train[, order(result, decreasing = TRUE)]
    n.pred.thr <- length(result[result >= threshold])
    n.pred <- min(n.pred, n.pred.thr)
    x.train <- x.train[, 1:n.pred]
    x.test <- newx[, colnames(x.train)]

    ## calculate parameters for OmicSense modeling
    coef <- data.frame(matrix(rep(NA, ncol(x.train) * (degree + 1)), ncol(x.train), degree + 1))
    sd <- rep(NA, ncol(x.train))
    for(i in 1:ncol(x.train)){
      reg <- lm(y.train ~ poly(x.train[, i], degree = degree, raw = TRUE))
      coef[i, ] <- summary(reg)$coefficients[, 1]
      sd[i] <- sqrt(mean(residuals(reg)^2))
    }

    ## prediction using training data
    p1 <- data.frame(matrix(rep(NA, nrow(x.train) * n.pred), nrow(x.train), n.pred))
    for(i in 1:nrow(x.train)){
      for(j in 1:n.pred){
        x2 <- 0
        for(k in 1:(degree + 1)){
          x2 <- x2 + coef[j, k] * (x.train[i, j] ^ (k - 1))
        }
        p1[i, j] <- x2
      }
    }
    p1 <- apply(p1, 1, mean, na.rm = TRUE)

    ## prediction using test data
    p2 <- data.frame(matrix(rep(NA, nrow(x.test) * n.pred), nrow(x.test), n.pred))
    for(i in 1:nrow(x.test)){
      for(j in 1:n.pred){
        x2 <- 0
        for(k in 1:(degree + 1)){
          x2 <- x2 + coef[j, k] * (x.test[i, j] ^ (k - 1))
        }
        p2[i, j] <- x2
      }
    }
    q.in <- apply(p2, 1, quant)

    mn <- min(y.train)
    mx <- max(y.train)
    dd <- 2
    l <- round(mn - (mx-mn)/2, digits = dd)
    u <- round(mx + (mx-mn)/2, digits = dd)
    range <- ((l * (10^dd)):(u * (10^dd)))/(10^dd)

    p2[p2 <= l | p2 >= u] <- NA
    stack <- array(rep(NA, nrow(x.test) * n.pred * length(range)), c(nrow(x.test), n.pred, length(range)))
    for(j in 1:n.pred){
      stack[, j, ] <- dnorm(rep(range, each = nrow(x.test)), p2[, j], sd[j])
    }
    stack[, apply(x.test, 2, sum) == 0, ] <- NA #missing values
    for(i in 1:nrow(x.test)){
      stack[i, -q.in[[i]], ] <- NA #outliers
    }
    stk <- apply(stack, c(1, 3), mean, na.rm = TRUE)

    p2 <- rep(0, nrow(x.test))
    for (i in 1:nrow(x.test)){
      p2[i] <- range[which.max(stk[i, ])]
    }

    pred.data <- adj(p1, y.train, p2)
    names(pred.data) <- row.names(x.test)
    pred.data
  }



  #sense.mix
  sense.mix <- function(x.train, y.train, newx, n.pred = n.pred, threshold = thresh){

    dim <- rep(NA, ncol(x.train))
    aic.rec <- rep(10^5, ncol(x.train))
    r.rec <- rep(0, ncol(x.train))
    for(i in 1:ncol(x.train)){
      for(j in 1:3){
        lm <- lm(y.train ~ poly(x.train[, i], degree = j, raw = TRUE))
        rss <- sqrt(sum((y.train - predict(lm, data.frame(x.train[, i])))^2))
        aic <- nrow(x.train)*(log(2 * pi * rss/nrow(x.train)) + 1) + 2 * (j + 2)
        r <- summary(lm)$r.squared
        if(aic.rec[i] > aic){
          dim[i] <- j
          aic.rec[i] <- aic
          r.rec[i] <- r
        }
      }
    }

    ord <- order(r.rec, decreasing = TRUE)
    x.train <- x.train[, ord]
    dim <- dim[ord]
    n.pred.thr <- length(r.rec[r.rec >= threshold])
    n.pred <- min(n.pred, n.pred.thr)
    x.train <- x.train[, 1:n.pred]
    x.test <- newx[, colnames(x.train)]

    coef <- data.frame(matrix(rep(NA, ncol(x.train) * 4), ncol(x.train), 4))
    sd <- rep(NA, ncol(x.train))
    for(i in 1:ncol(x.train)){
      reg <- lm(y.train ~ poly(x.train[, i], degree = dim[i], raw = TRUE))
      coef[i, 1:(dim[i] + 1)] <- summary(reg)$coefficients[, 1]
      sd[i] <- sqrt(mean(residuals(reg)^2))
    }

    p1 <- data.frame(matrix(rep(NA, nrow(x.train) * n.pred), nrow(x.train), n.pred))
    for(i in 1:nrow(x.train)){
      for(j in 1:n.pred){
        x2 <- 0
        for(k in 1:(dim[j] + 1)){
          x2 <- x2 + coef[j, k] * (x.train[i, j] ^ (k - 1))
        }
        p1[i, j] <- x2
      }
    }
    p1 <- apply(p1, 1, mean, na.rm = TRUE)

    p2 <- data.frame(matrix(rep(NA, nrow(x.test) * n.pred), nrow(x.test), n.pred))
    for(i in 1:nrow(x.test)){
      for(j in 1:n.pred){
        x2 <- 0
        for(k in 1:(dim[j] + 1)){
          x2 <- x2 + coef[j, k] * (x.test[i, j] ^ (k - 1))
        }
        p2[i, j] <- x2
      }
    }
    q.in <- apply(p2, 1, quant)

    mn <- min(y.train)
    mx <- max(y.train)
    dd <- 2
    l <- round(mn - (mx-mn)/2, digits = dd)
    u <- round(mx + (mx-mn)/2, digits = dd)
    range <- ((l * (10^dd)):(u * (10^dd)))/(10^dd)

    p2[p2 <= l | p2 >= u] <- NA
    stack <- array(rep(NA, nrow(x.test) * n.pred * length(range)), c(nrow(x.test), n.pred, length(range)))
    for(j in 1:n.pred){
      stack[, j, ] <- dnorm(rep(range, each = nrow(x.test)), p2[, j], sd[j])
    }
    stack[, apply(x.test, 2, sum) == 0, ] <- NA #missing values
    for(i in 1:nrow(x.test)){
      stack[i, -q.in[[i]], ] <- NA #outliers
    }
    stk <- apply(stack, c(1, 3), mean, na.rm = TRUE)

    p2 <- rep(0, nrow(x.test))
    for (i in 1:nrow(x.test)){
      p2[i] <- range[which.max(stk[i, ])]
    }

    pred.data <- adj(p1, y.train, p2)
    names(pred.data) <- row.names(x.test)
    pred.data
  }



  #sense.gauss
  sense.gauss <- function(x.train, y.train, newx, n.pred = n.pred){

    result <- rep(NA, ncol(x.train))
    for(i in 1:ncol(x.train)){
      result[i] <- summary(lm(y.train ~ poly(x.train[, i], degree = 3, raw = TRUE)))$r.squared
    }

    x.train <- x.train[, order(result, decreasing = TRUE)]
    x.train <- x.train[, 1:n.pred]
    x.test <- newx[, colnames(x.train)]

    mn <- min(y.train)
    mx <- max(y.train)
    dd <- 2 #decimal digits
    l <- round(mn - (mx-mn)/2, digits = dd)
    u <- round(mx + (mx-mn)/2, digits = dd)
    range <- ((l * (10^dd)):(u * (10^dd)))/(10^dd)

    stack1 <- array(rep(NA, nrow(x.train) * n.pred), c(nrow(x.train), n.pred))
    stack2 <- array(rep(NA, nrow(x.test) * n.pred * length(range)),
                    c(nrow(x.test), n.pred, length(range)))
    y.est <- array(rep(NA, nrow(x.test) * n.pred), c(nrow(x.test), n.pred))
    for(j in 1:n.pred){
      invisible(capture.output(fit <- gausspr(x.train[, j], y.train, variance.model = TRUE)))
      stack1[, j] <- predict(fit, x.train[, j])
      y.val <- predict(fit, x.test[, j])
      y.val[y.val <= l | y.val >= u] <- NA
      y.est[, j] <- y.val
      sd.val <- predict(fit, x.test[, j], type = "sdeviation")
      for(i in 1:nrow(x.test)){
        stack2[i, j, ] <- dnorm(range, y.val[i], sd.val[i])
      }
    }
    q.in <- apply(y.est, 1, quant)

    p1 <- apply(stack1, 1, mean)
    stack2[, apply(x.test, 2, sum) == 0, ] <- NA #missing values
    for(i in 1:nrow(x.test)){
      stack2[i, -q.in[[i]], ] <- NA #outliers
    }
    stk <- apply(stack2, c(1, 3), mean, na.rm = TRUE)

    p2 <- rep(0, nrow(x.test))
    for (i in 1:nrow(x.test)){
      p2[i] <- range[which.max(stk[i, ])]
    }

    pred.data <- adj(p1, y.train, p2)
    names(pred.data) <- row.names(x.test)
    pred.data
  }

  #OmicSense function
  if(degree > 0){
    pred.data <- sense.poly(x, y, newx, method = method, n.pred = n.pred, threshold = thresh)
  }else if(degree == 0){
    pred.data <- sense.mix(x, y, newx, n.pred = n.pred, threshold = thresh)
  }else{
    pred.data <- sense.gauss(x, y, newx, n.pred = n.pred)
  }
  pred.data #return
}


